### DISCUSSION TITLE
Time Signature

### DISCUSSION THREAD

**Discussion by NoAliby** [hydrogen-music/hydrogen#1680] [Q&A] [answered]
I have created a full song in a 3/4 time.
When I export the Midi I cannot find where to incorporate any time signature. 
So when I play it as a drum loop it defaults to 4/4 and so does not function correctly.
Any tips?

**Comment by NoAliby** [comment, marked as answer]
As a work around I used "MidiQuickFix" to edit the .mi file directly.
I delete the M:TimeSignature event line and insert the same event again with it set as 3/4
https://sourceforge.net/projects/midiquickfix/

**Reply by theGreatWhiteShark** [reply]
Thanks for the hint!

Hydrogen itself does not support time signatures. In recent version the dialog used to set the length of a measure was redesigned to represent something similar to a time signature. But internally it is totally agnostic to such a concept and it (alongside a lot of other newer features) hasn't been introduced into the MIDI export yet.

**Reply by NoAliby** [reply]
@theGreatWhiteShark it would be helpful if we could enter a desired time signature for a Song so it can be used in Midi export (and the LilyPond export, which I know is also experimental).

I agree with you that while initially to default to or have a check box to use the size #/#  seems logical, it is probably inappropriate 90% of the time as the export is effectively from the "Song Mode". Also the size in "Pattern Mode" is purely what is convenient or required to construct or complete a bar or phrased rhythm section. However, the Metronome currently uses the #/# for PATTERNS, which challenges this thinking somewhat. (In SONG mode the metronome will use that per pattern, when multiple patterns are played it will scale to the max length and then use that for that section, i.e. the click track can be changing to different 1-beats counts throughout the song).

Here are some ideas.
1. The least impact on others could be to add it to the Export prompt for Midi and LilyPond (it should remember the last used value)
2. There could be a field in Song Properties we could populate with the time signature (be used anything needing it)
3. If it is in the Transport toolbar it could also be used for the Metronome and the two exports. It would be very helpful with playback of non-4/4 songs, but it introduces a number of questions to be addressed (e.g. whether the 1-click should always be on "beat 1" when the Pattern loops, but count using the "time-signature" when the Song loops, not to mention about how people are using it and if this will "break" their workflow?)
4. Another discrete (advanced user) option would be to be able to change the default in Midi preferences (4/4 is hard coded at the moment)
 
Option 1 (or 4) may have the least or no impact on anyone else.

**Reply by theGreatWhiteShark** [reply]
Hey @NoAliby ,

Hmmm... I don't really like the idea of hard coding the signature in the preferences, properties, or export dialog. It will rot very quickly and is not able to represent songs with changing signature.

Shouldn't the time signature be determined by the length of the longest pattern in a column divided by the length of 4 quarters? If so, we would already know the signature of each column and could add it to the first note of the corresponding column in the MIDI file. No additional user interaction required and no chance the signature gets incorrect when editing the song.

But I have not looked into the MIDI standard yet. (Did you by any chance?) I think the approach above would be feasible but it's just a guess.

**Reply by NoAliby** [reply]
@theGreatWhiteShark I have attached a screen shot of Hydrogen with a song, together with another program that shows the EVENT details of the exported MIDI file.

At the moment Hydrogen exports **All songs** to MIDI with a Time Signature of **4/4**. It is a fixed value.
You can see that in the header EVENTS details of the Midi file (Yellow Highlight). Somewhere in the code it must have 4/4 entered in the export command. The midi file exported from Hydrogen only has one Time Signature EVENT. This Time Signature applies to the whole Midi song, which is typical for midi files and exceptions are rare.

At the moment the MIDI export command is getting the TEMPO from the Hydrogen song (Red markup).
If there was a way to get a user input to replace the hard coded 4/4 that would be great. (e.g. Light Green markup)
![Hydrogen - Midi Time Signature](https://github.com/user-attachments/assets/35bcecbb-3fef-4b4d-aff1-58c8b7bc2cdb)

_(As noted above at I currently open every MIDI file and manually overwrite the TIME SIGNATURE, my drum machine/pedal uses that to sync fills and transitions between parts. If Hydrogen could do it it would save a lot of time - fortunately most songs are 4/4 so the hard code work 95% of the time)_

**Reply by theGreatWhiteShark** [reply]
> together with another program that shows the EVENT details of the exported MIDI file.

which one did you use?

> At the moment the MIDI export command is getting the TEMPO from the Hydrogen song

Yes. But supporting tempo changes throughout the song is still an open ticket (see #933). While tackling this I could also implement a way to derive the time signature from the pattern's length.

How come you want to set 6/8 as time signature while song size is set to 12/8? Was this done intentionally?

**Reply by NoAliby** [reply]
> which one did you use?

I use [MidiQuickFix](https://github.com/hydrogen-music/hydrogen/discussions/1680#discussioncomment-4280352) as noted earlier in the thread, as it enables easy access to EVENTS without the bloat of a fully fledged sequencer.

> While tackling this I could also implement a way to derive the time signature from the pattern's length.

Deriving the Time Signature from the pattern length would not help me with my use of Hydrogen in programming my [BeatBuddy](https://forum.singularsound.com/t/beatbuddy-drum-kit-for-hydrogen/27438), (unless there was a way of manually overriding that). 

1. In BeatBuddy a drum **Fill** (within a loop part), or a drum **Transition** (when moving to a new loop part) requires the **first** beat of the next measure - it uses this to sync with the next measure, regardless of whether you trigger the fill or transition exactly at the start of the playing loop's measure or slightly delayed. That means the fill or transition for a 6/8 loop must have a pattern length of 7/8.

> How come you want to set 6/8 as time signature while song size is set to 12/8? Was this done intentionally?

2. I make each loop part a single Hydrogen **pattern** for ease of transcribing live drumming and for exporting, which means if a 6/8 loop has two measures the pattern length in Hydrogen would be set to 12/8

**Reply by theGreatWhiteShark** [reply]
Ah, I see. You want to manually overwrite the time signature with incorrect values to cope with limitations of an external application/device.

> In BeatBuddy a drum Fill (within a loop part), or a drum Transition (when moving to a new loop part) requires the first beat of the next measure - it uses this to sync with the next measure, regardless of whether you trigger the fill or transition exactly at the start of the playing loop's measure or slightly delayed. That means the fill or transition for a 6/8 loop must have a pattern length of 7/8.

Would programming the pattern in 6/8 and then extending it by an empty space to 7/8 yield the same results?

**Reply by theGreatWhiteShark** [reply]
Hey @NoAliby ,

You are on Linux, right? Could you check whether MIDI export works fine in [this](https://ci.appveyor.com/api/buildjobs/1h3i8u6cm26lhobs/artifacts/Hydrogen-2.0.0-pre-alpha-9ebc36d79-x86_64.AppImage) version of Hydrogen?

But you should backup your `$HOME/.hydrogen/data` folder beforehand.

**Reply by NoAliby** [reply]
> You are on Linux, right? Could you check whether MIDI export works fine in Hydrogen-2.0.0-pre-alpha version of Hydrogen?

I haven't run my Linux box for a while, but will see if I can test it.

**Reply by theGreatWhiteShark** [reply]
You could also use the [macOS](https://ci.appveyor.com/api/buildjobs/xnd90jwj9w2kotly/artifacts/Hydrogen-2.0.0-pre-alpha-8c0b6fdee.dmg) or [Windows](https://ci.appveyor.com/api/buildjobs/oxix2oa2vnobymmd/artifacts/Hydrogen-2.0.0-pre-alpha-8c0b6fdee-win64.exe) version.

**Comment by NoAliby** [comment]
Hi @theGreatWhiteShark ,
That works as you described where it takes the length of loop and the subdivision.
_(Screenshots show Hydrogen in the background with the MidiQuickFix window showing the time-signature in the Midi file.)_

This appears to work (6/8) with conventional loops (patterns) that are only one measure or bar long. From what I understand this will be the same as having a series of multiple patterns, where the "prediction" uses the size and subdivision of the uppermost pattern.
![Screenshot 2025-03-22 172613](https://github.com/user-attachments/assets/34f5d48b-5a12-45f5-8d31-c913c5c38137)

It does not work (12/8 instead of 6/8) if someone uses a double measure in the pattern - a size of 12, which often could makes sense as many drum loops have two measures that are identical except for emphasis around the last beat (e.g. open high-hat)
![Screenshot 2025-03-22 171800](https://github.com/user-attachments/assets/82305172-cfc2-44c4-9bda-bc63c732ebe4)

If you have any "technical" or special case loops, such as what I need for the [BeatBuddy](https://www.singularsound.com/products/beatbuddy) loops, then this also needs editing in an external program like **MidiQuickFix**. (predicts 7/8 instead of 6/8)
![Screenshot 2025-03-22 172649](https://github.com/user-attachments/assets/b4875c34-aca5-472c-a6c5-c453cec40a2f)

My perspective:
1. Using the size / subdivision as the basis for the time-signature is a better default than 4/4. The manual should explain that the size / subdivision of the first pattern (the first pattern in order of the song might be logical, but the first pattern in the list, i.e. top, would enable you to have a "first pattern" just being 2 clicks).
2. To be useful for ALL ways people are using Hydrogen there should be an option in the export window to change the predicted Time-Signature - this would be fantastic. (See my mock-up of the export window)
![Hydrogen Midi Export WIndows](https://github.com/user-attachments/assets/b0b360eb-4955-4c4b-abc2-38675e101adb)

**Reply by theGreatWhiteShark** [reply]
> From what I understand this will be the same as having a series of multiple patterns, where the "prediction" uses the size and subdivision of the uppermost pattern.

If you have several patterns activated in a single column, it will take the size and subdivision of longest one. Regardless of its position. This applies to each column separately. A column without any pattern activated will fall back to 4/4.

> It does not work (12/8 instead of 6/8) if someone uses a double measure in the pattern - a size of 12, which often could makes sense as many drum loops have two measures that are identical except for emphasis around the last beat (e.g. open high-hat)

If they are identical except for an emphasis, you could also put the common part into a dedicated pattern, the emphasis into individual patterns, and use several patterns per column.

![2025-03-23-151012_3840x1220_scrot](https://github.com/user-attachments/assets/ce33b60c-ab70-4011-8994-07ed5fe92e88)

This should yield the same results as the pattern you programmed and you would not need to do changes in the common part twice.

> If you have any "technical" or special case loops, such as what I need for the [BeatBuddy](https://www.singularsound.com/products/beatbuddy) loops, then this also needs editing in an external program like MidiQuickFix. (predicts 7/8 instead of 6/8)

Well, having to intentionally set an incorrect time signature to work around a bug (?) of an external application is indeed a special case. MIDI export is always done for the whole song. Would it help to add an empty 1/8 after the 6/8 pattern?

![2025-03-23-151145_3840x1220_scrot](https://github.com/user-attachments/assets/c075314d-0f49-4835-9b4e-72edc7e3722d)

What does the BeatBuddy needs the time signature for if it's not able to handle it properly?

> To be useful for ALL ways people are using Hydrogen there should be an option in the export window to change the predicted Time-Signature - this would be fantastic. (See my mock-up of the export window)

It's about being useful to as many people as possible. There might be conflicting use cases that can not be supported at the same time and even if this is not the case, an overly complex UX supporting all possible ways is often still less useful. Some people won't understand and use the feature while they would be very happy with a simplified version.

The problem with the mock up is that it only covers a very specific use case of either a single pattern in the song or a constant time signature throughout the whole song. The feature we are talking about is exporting the whole song into a MIDI file. If we would allow manual overwrite of the time signature, we would need to allow this for all columns and provide a table with each time signature per column. For large songs with a lot of pattern this will quickly escalate. And why only allowing manual overwrite of time signatures and not just tempo and markers? For the average user such a table would only cause confusion.

Maybe such a feature as in your mock up would be a perfect candidate for a fork. After all it only addresses workarounds for a specific revision of the BeatBuddy.

**Reply by NoAliby** [reply]
From a music theory side, a full song should only have **one time signature** for the entire song. Multiple columns and multiple patterns - whether sequential or simultaneous, regardless of individual sizes - will all combine into one song, and it will still just have a single time-signature. This appears only once on the music score. In the same way only once in the midi file.
![Drum Music](https://github.com/user-attachments/assets/d803182c-fe1e-4ac0-820b-0de904803957)

**Reply by NoAliby** [reply]
> What does the BeatBuddy needs the time signature for if it's not able to handle it properly?

Like any sequencer, BeatBuddy needs the time signature for any Midi drum loop, so it knows when a **beat one** occurs.

> Would it help to add an empty 1/8 after the 6/8 pattern?

That is an idea. For a 6/8 song I could create the size 7 with a subdivision of 8 by using one size 6, div 8 pattern followed by a single 1 size, div 8 pattern and then join the two in a song and export it to midi. With the v2.0_alpha code, would that predicted the time signature as 6/8?

**Reply by theGreatWhiteShark** [reply]
> From a music theory side, a full song should only have one time signature for the entire song. Multiple columns and multiple patterns - whether sequential or simultaneous, regardless of individual sizes - will all combine into one song, and it will still just have a single time-signature. This appears only once on the music score. In the same way only once in the midi file.

Hmm. It's been a while since I had music classes but I'm pretty sure I have seen a couple of signature changes in sheet music. Wikipedia even has dedicated paragraphs in the Time Signature page regarding this matter: https://en.wikipedia.org/wiki/Time_signature#Mixed_meters and https://en.wikipedia.org/wiki/Time_signature#Irregular_bar

As for MIDI, the [Standard MIDI File Specification](https://midi.org/standard-midi-files) explicitly states

> As with Set Tempo changes, the Time Signature can change throughout an SMF [...]

(right at the bottom of the document)

**Reply by theGreatWhiteShark** [reply]
> With the v2.0_alpha code, would that predicted the time signature as 6/8?

Yes. But, in addition, it would add another time signature of 1/8 after passing the first pattern.

**Reply by NoAliby** [reply]
You are correct that you can change time signature, but it is not common. When it does it would be for an entire chorus or bridge not a single measure or bar. 
Adding a 1/8 would make it disfunctional as the metronome would loose the 6/8 rhythm and go...
**1** 2 3 4 5 6,   **1**,   **1** 2 3 4 5 6 ...

**Reply by theGreatWhiteShark** [reply]
> When it does it would be for an entire chorus or bridge not a single measure or bar.

Hydrogen should only add a time signature event in case the size of the longest pattern in a column differs compared to the previous one. If it's all the same size for the entire chorus/song, there won't be an additional time signature event.

> Adding a 1/8 would make it disfunctional as the metronome would loose the 6/8 rhythm and go...
1 2 3 4 5 6, 1, 1 2 3 4 5 6 ...

And that's a problem? I still don't fully understand the problem with your BeatBuddy. 

You want to program a pattern like this

1 2 3 4 5 6 7, 1 2...

but the trick the BeatBuddy into thinking it is 

1 2 3 4 5 6, 1 2...

instead? But by having 7 ticks instead of 6 in the pattern, you also move the end of the track encoded in the MIDI file. All following MIDI notes will be off one tick. E.g. if you ever want your pattern to be the basis of a more complex/longer one and activate another one in the second column, The click track will be off as well.

If it is just about the click within the BeatBuddy, you may be better off with having a dedicated click track - pattern you activate in the same column.

**Comment by NoAliby** [comment]
Can you start a topic where you can share changes and comments or bugs in _Hydrogen-2.0.0-pre-alpha_?
- I like the way you have presented the hit-hat open.
- I could not get sound playing. I tried to change the Midi Out and the Audio settings but no sound.

**Reply by theGreatWhiteShark** [reply]
> Can you start a topic where you can share changes

Good point. I added [this page](https://github.com/hydrogen-music/hydrogen/wiki/What's-new-in-version-2.0%3F-How-to-migrate%3F) to our wiki. Is it useful?

>  you can share [...] bugs in Hydrogen-2.0.0-pre-alpha?

You can open an new issue. This is makes book keeping a lot easier than having threads in discussions.

> I could not get sound playing. I tried to change the Midi Out and the Audio settings but no sound.

Could you start Hydrogen with `hydrogen -VDebug` and post the log messages?

**Reply by NoAliby** [reply]
Thanks @theGreatWhiteShark, the V2 summary page is very useful.
I will check the V2 installation again on my other machine and let you know. If it dies not work I will add a new ISSUE, use the `- VDebug` option and include the files.
